### ISSUE TITLE
Additional arguments for setcolorder: before, after

### ISSUE THREAD

**Issue by Plebejer** [Rdatatable/data.table#4358]
It would be nice to have two additional arguments (`before = NULL` and `after = NULL`) in `setcolorder()` for reordering one or more columns to a position other than the "front" of the `data.table`. This would make it possible to place a single column after the, say, 45th column, without having to list all these 46 columns.

After having a look at the code, I think I could try to file a PR if you think this feature is helpful.

**Comment by Atrebas** [comment]
Was just thinking about it. It has been implemented in dplyr 1.0.0: [`relocate()`](https://dplyr.tidyverse.org/reference/relocate.html). Would be good to have it in data.table as well.

**Comment by dracodoc** [comment]
I had some implementation suggested on #4668. If there is a plan to add this feature, I'd love to submit a PR.

**Comment by Atrebas** [comment]
Thanks for the PR. As a humble data.table user, I think I would prefer having `before` and `after` parameter than `anchor`. Just my two cents.

**Comment by ben-schwen** [comment]
Changing `setcolorder` like below should do the trick. This also ensures backward compatibility.

```
setcolorder = function(x, neworder=key(x), before=NULL, after=NULL)
{
  if (is.character(neworder) && anyDuplicated(names(x)))
    stop("x has some duplicated column name(s): ", paste(names(x)[duplicated(names(x))], collapse=","), ". Please remove or rename the duplicate(s) and try again.")
  # if (!is.data.table(x)) stop("x is not a data.table")

  if (!is.null(before) && !is.null(after))
    stop("Provide either before= or after= or none but not both.")

  neworder = colnamesInt(x, neworder, check_dups=FALSE)  # dups are now checked inside Csetcolorder below

  if (!is.null(before))
    neworder = c(setdiff(seq_len(colnamesInt(x, before) - 1), neworder), neworder)

  if (!is.null(after))
    neworder = c(setdiff(seq_len(colnamesInt(x, after)), neworder), neworder)

  if (length(neworder) != length(x)) {
    #if shorter than length(x), pad by the missing
    #  elements (checks below will catch other mistakes)
    neworder = c(neworder, setdiff(seq_along(x), neworder))
  }
  .Call(Csetcolorder, x, neworder)
  invisible(x)
}
```

**Comment by dracodoc** [comment]
Thanks for all the comments!

@ben-schwen After reading your code I found all my code can be replaced with yours. Mine was originated from a separate function, the original design is simple and easy to understand. To make it work with new api (instead of `before, anchor`, change to `before=anchor`) I spent too much lines on parameter checking.

Your code is much more shorter because
- `colnamesInt` can convert name to index and keep index, I didn't know about it
- seq_len replaced my creation of slice to deal with 1:0 case
- reusing the padding part of original function removed the need of right side part

This is a much better implementation, fully utilized existing structures.

I don't think I need to edit my PR anymore, can you create a PR with your code?

(The only suggestion I want to make is the error message could be "before= and after= cannot be specified at the same time", this is subjective but I feel it be easier to read)

**Comment by dracodoc** [comment]
Seeing @ben-schwen  may not have time to create the PR, I created a PR  #4691 with his implementation and full credit go to @ben-schwen .

**Comment by Feakster** [comment]
What became of this in the end?

**Comment by matthiasgomolka** [comment]
@dracodoc, your PR does not pass the build pipeline. Are you willing to give it another try? If not, I will try to find the time to take over.

**Comment by dracodoc** [comment]
I'm not sure what was the problem. Was it just test case orders? I just modified the test cases and pushed an update, let's see if it pass this time. The tests run without problem for me.

**Comment by matthiasgomolka** [comment]
I don't know. I can't look into the detailed test results.

**Comment by dracodoc** [comment]
I think I knew what caused original error. I'm updating the commit and wait for next build check.

**Comment by dracodoc** [comment]
Looks all checks passed this time. @matthiasgomolka can you have a look?

**Comment by matthiasgomolka** [comment]
Yes, now it passes the tests. But I don't know the usual procedure if and when a new PR is merged.

@jangorecki, you already had a look at this issue. Can you explain the next steps?

**Comment by jangorecki** [comment]
Once PR is passing tests the next step is to wait for reviews and merging.
